### PR TITLE
Configure Controller, Integration, Mailer, and View test HTML assertions

### PR DESCRIPTION
**Pull request by seanpdoyle**
Related to https://github.com/rails/rails/pull/55743
This is a re-submission of https://github.com/rails/rails/pull/41291.

The problem
===

Framework test cases that render HTML also include a module of
assertions provided by the [rails-dom-testing][] gem. These test cases
include:

* [ActionController::TestCase][] (through [ActionController::TestCase::Behavior][])
* [ActionDispatch::IntegrationTest][] (through [ActionDispatch::Assertions][])
* [ActionMailer::TestCase][] (through [ActionMailer::TestCase::Behavior][])
* [ActionView::TestCase][] (through [ActionView::TestCase::Behavior][])

Once included, additional user-provided test assertions can collide with
the `Rails::Dom::Testing::Assertions` methods.

For example, the `assert_select` method from
`Rails::Dom::Testing::Assertions` (asserts using a CSS selector)
directly conflicts with the [assert_select][] method from
`Capybara::Minitest::Assertions` (asserts the presence of a `<select>`
HTML element), which could be included by a user into one of the test
case classes mentioned above.

While these collisions can be handled on a case-by-case basis, it isn't
possible to opt-out of `Rails::Dom::Testing::Assertions` support at the
application level.

The proposal
===

This commit introduces framework-specific configuration values:

* `config.action_controller.dom_testing_assertions` to configure [ActionController::TestCase][]
* `config.action_dispatch.dom_testing_assertions` to configure [ActionDispatch::IntegrationTest][]
* `config.action_mailer.dom_testing_assertions` to configure [ActionMailer::TestCase][]
* `config.action_view.dom_testing_assertions` to configure [ActionView::TestCase][]

When set to `:rails_dom_testing`, those tests include a new
framework-private `ActionView::RailsDomTestingAssertions` module to
preserve the existing behavior:

```ruby
# config/application.rb
config.action_view.dom_testing_assertions = :rails_dom_testing

class RailsDomTestingViewTest < ActionView::TestCase
  test "can use rails-dom-testing assertions" do
    render html: <<~HTML
      <label for="textbox">A textbox</label>
      <input id="textbox">
    HTML

    assert_select "#textbox"    # assert with a CSS id selector
    assert_select "input"       # assert with a CSS class selector
  end
end
```

**Each framework defaults to `:rails_dom_testing`**.

When set to `:none`, no assertions are included, providing applications
with an opportunity to include whichever methods they need to, or
customize on per-test basis.

```ruby
# config/application.rb
config.action_view.dom_testing_assertions = :none

class CustomViewTest < ActionView::TestCase
  include Capybara::Minitest::Assertions

  test "can use custom assertions" do
    render html: <<~HTML
      <label>
        A select
        <select></select>
      </label>
    HTML

    assert_select "A select"    # assert with <label> text
  end

  def page
    Capybara.string(rendered)
  end
end
```

[rails-dom-testing]: https://github.com/rails/rails-dom-testing
[ActionController::TestCase]: https://edgeapi.rubyonrails.org/classes/ActionController/TestCase.html
[ActionController::TestCase::Behavior]: https://edgeapi.rubyonrails.org/classes/ActionController/TestCase/Behavior.html
[ActionDispatch::IntegrationTest]: https://edgeapi.rubyonrails.org/classes/ActionDispatch/IntegrationTest.html
[ActionDispatch::Assertions]: https://edgeapi.rubyonrails.org/classes/ActionDispatch/Assertions.html
[ActionMailer::TestCase]: https://edgeapi.rubyonrails.org/classes/ActionMailer/TestCase.html
[ActionMailer::TestCase::Behavior]: https://edgeapi.rubyonrails.org/classes/ActionMailer/TestCase/Behavior.html
[ActionView::TestCase]: https://edgeapi.rubyonrails.org/classes/ActionView/TestCase.html
[ActionView::TestCase::Behavior]: https://edgeapi.rubyonrails.org/classes/ActionView/TestCase/Behavior.html
[assert_select]: https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Minitest/Assertions:assert_select

### CODE REVIEW

**Comment by nimmolo** [comment]
@seanpdoyle maybe the lack of context is the reason for the lack of further discussion on this PR. It's such a useful implementation... and it seems to me you've made requested modifications and done the benchmarking. 

Would it make sense to abandon this newer PR, and consolidate these changes into the [older PR branch](https://github.com/rails/rails/pull/41291), so reviewers and core committers have the reminders of the work and review that has already gone into this?

**Comment by cpjmcquillan** [comment]
@seanpdoyle awesome work, +1-ing that it would be great if this was reconsidered. 

Oh, and thanks again for all your fantastic contributions to the community!

**Comment by nimmolo** [comment]
thanks @p8, @cpjmcquillan and @seanpdoyle !

**Comment by seanpdoyle** [comment]
Unfortunately, https://github.com/rails/rails/pull/41291 was closed due to inactivity, and as a result, the context for this discussion must be split between two pull requests.

Due to that split, I need to ping @eileencodes @rafaelfranca @zzak, since you were all involved in the original discussion.

Given [the concerns about the performance impacts of the original changeset](https://github.com/rails/rails/pull/41291#issuecomment-772556309), I've re-structured the proposal to a point where it's entirely opt-in.

Is this approach more viable than the original?

**Comment by rafaelfranca** [comment]
I think so, but I also think is more confusing. Why would someone chose this new list of assertions over the existing ones? Rails is opinionated, and having two set of assertions for the same kind of tests seems that we don't have an opinion in the matter.

I still don't understand why we wouldn't improve rails-dom-testing. If the capybara assertions were not slower I'd even say we could just stop using rails-dom-testing, but that doesn't seems to be the case since those capybara assertions are almost 2 times slower in the worse case.

**Comment by seanpdoyle** [comment]
@rafaelfranca thank you for jumping back into this conversation! I'm sorry for the long pause and the change in venue.

> Why would someone chose this new list of assertions over the existing ones? 

Capybara's built-in selectors are more robust and flexible than what `rails-dom-testing` provides. Built-in selectors like `:button` or `:field` are more sophisticated than their String selector `rails-dom-testing` counterparts. For example, Capybara can find a `<button>` element based on its text content, `[type]` value, and presence or absence of the `[disabled]` attribute:

```ruby
# <button>Submit</button>

assert_button "Submit"
assert page.find(:button, "Submit")

# <button type="reset">Reset</button>

assert_button type: "reset"
assert page.find(:button, type: "reset")

# <button disabled>Submit</button>
assert_no_button "Submit"
page.find(:button, "Submit") #=> raises!

assert_button "Submit", disabled: true
assert page.find(:button, disabled: true)

# <button>Submit</button>
# <button type="button">Cancel</button>

assert_button { _1.text.include?("Submit") && _1["type"] != "button" }
```

`rails-dom-testing` lacks built-in abstractions, and instead relies on String selectors. This leaves the burden of responsibility with applications to develop their own abstractions and helpers. Without those abstractions, assertions about `<button>` elements might resemble:

```ruby
# <button>Submit</button>

assert_select "button", "Submit"

# <button type="reset">Reset</button>

assert_select 'button[type="reset"]'

# <button disabled>Submit</button>
assert_no_select "button:not([disabled])", "Submit"

assert_select "button[disabled]", "Submit"
```

Some selectors can be mapped directly one-to-one from Capybara to `rails-dom-testing`. However, it gets tricky for elements that derive their "text" from somewhere other than the element's `.textContent`. Take, for example, an `<input type="submit">` or `<input type="button">`. Capybara's built-in selectors are capable of distinguishing between the two:

```ruby
# <input type="submit" value="Submit">
# <input type="button" value="Submit">
# <button>Submit</button>
# <button type="button">Submit</button>

assert_button "Submit", count: 4
assert_button "Submit", type: "button", count: 2
```

Calls to `assert_select`, however, must account for the `textContent`-`[value]` differences _and_ the `input`-`button` differences. While it's true that applications or `rails-dom-testing` itself could come up with abstractions to paper over those differences, that effort would duplicate what Capybara already provides.

Another feature of Capybara's selectors that would be challenging to recreate in `rails-dom-testing` is its support for locating `:field` (and `:fillable_field`, and `:file`) based on corresponding `<label>` text or `[aria-label]` value.

```ruby
# <label for="a_text_field">A text field</label>
# <input id="a_text_field" value="an initial value">

assert_field "A text field", with: "an initial value"
```

The accessibility benefits alone feel valuable enough to warrant a migration. Personally, I find support for asserting that form controls are properly labelled without the need for JavaScript-enabled System Tests very compelling.

> having two set of assertions for the same kind of tests seems that we don't have an opinion in the matter.

I think we're in agreement here. My perspective is that since **most applications already depend on Capybara for driving their System Tests**, integrating those predictable and robust selectors into their Rack Test driven integration tests could *halve their maintenance burden*. Choosing Capybara over `rails-dom-testing` means they only need to create the abstraction once, and use it anywhere their test harness interacts with HTML. Some of those applications might already have their own suite of domain-specific bespoke Capybara selectors (or depend on third-party selectors like those provided by [capybara_accessible_selectors](http://github.com/citizensadvice/capybara_accessible_selectors/)). Should they also maintain a collection of bespoke `rails-dom-testing` selectors, helpers, and abstractions for their Integration test suite?

>  If the capybara assertions were not slower I'd even say we could just stop using rails-dom-testing, but that doesn't seems to be the case since those capybara assertions are almost 2 times slower in the worse case.

This is an unfortunate concern. My original proposal considered replacing `rails-dom-testing`, but felt that would be too disruptive. I wonder what kind of effort it would require to make Capybara a more performant option. From my perspective, I consider optimizing Capybara a more cost-effective use of time and effort than reverse engineering some of Capybara's features back into `rails-dom-testing`.

**Comment by rafaelfranca** [comment]
To me, given all options capybara helpers provide over our own helpers I can't see why Rails should not provide that as default, other than the performance concerns. We spent a considerable amount of time trying to make `ActionDispatch::IntegrationTest` as fast as `ActionController::TestCase` to open the avenue of deprecating/removing the latter, so making capybara helpers the default in `IntegrationTest` would goes against that.

But, if we can make Capybara helpers as fast as the ones we have today, I don't see a reason why we should not do that. I totally agree that if not too hard, we should spend time optimizing Capybara instead of reimplementing it.

**Comment by eileencodes** [comment]
> My perspective is that since most applications already depend on Capybara for driving their System Tests

I'm not sure this is a true statement. I can see why it seems that way, but any apps upgrading wouldn't be forced to add Capybara unless they wanted system tests and any application that's generated as just an API skips system tests. It's also pretty easy to remove. So while we can say a fair amount of applications already depend on Capybara it's much less of a requirement than replacing rails-dom-testing with Capybara would be. We'd be making a hard requirement on Capybara for default integration testing, whereas right now it's only a hard requirement for system testing.

> But, if we can make Capybara helpers as fast as the ones we have today, I don't see a reason why we should not do that. 

I think this is a big concern of mine as well, since I worked on making integration tests faster so they were in line with functional/controller tests. I wouldn't be supportive of undoing that work.

The other side of it is that currently we control rails-dom-testing but we have no control over how and when Capybara fixes bugs or adds features. I'm sure that it's well maintained but it is a complex codebase and would make it harder for the core team to have influence into how it affects the framework and testing environments. There is a non-zero chance it would increase the maintenance burden for us. This is fine for system testing because there aren't many other alternatives we could plug and play into Rails, and system testing isn't your default go to when testing the majority of your framework (mainly because system testing is pretty slow and should be mainly used for testing how Rails and JS interact).

I can tell you from first hand experience that since I'm not that familiar with Capybara it's been difficult for me to decide when to merge changes into Rails that "fix" system tests. I'm often relying on the PR author to have done due diligence into the right fix because I don't know Capybara well enough.

TL;DR:

My main concerns for moving to Capybara from rails-dom-testing are maintenance burden and performance. Improving Capybara perf gets rid of one my concerns.

**Comment by seanpdoyle** [comment]
Thank you two for your perspectives. While I'm personally interested in exploring ways to optimize Capybara, I'm curious about the viability of this Pull Request in its current form.

The Capybara assertions are opt-in, since the `ActionDispatch::Assertions::CapybaraAssertions` module isn't mixed into `ActionDispatch::IntegrationTest` by default.

This changeset isn't replacing or deprecating `rails-dom-testing`. It also isn't _introducing_ a Capybara dependency where one did not exist before (right?). Does the file's presence mean that it's autoloaded in a way that carries along the Capybara dependency from the inline `gem "capybara"` line?

This diff is the least disruptive version of these changes I could imagine. If the team isn't interested in accepting them, could we instead determine a more public interface so that a third-party gem could extend `ActionDispatch::IntegrationTest` without overriding the very private `_mock_session` method?

**Comment by rafaelfranca** [comment]
> could we instead determine a more public interface so that a third-party gem could extend `ActionDispatch::IntegrationTest` without overriding the very private `_mock_session` method?

I'm up for that.

While I can see this PR as a good compromise it goes against the opinionated nature of Rails. We are including 2 solutions inside the framework, when we usually are very opinionated on what should be used. An API to allow people to build on top of our system gives people to flexibility to chose, while keeping our opinion.

**Comment by jorgemanrubia** [comment]
I would personally LOVE to use capybara assertions in regular integration tests. I think this is a great addition. What do you think @dhh @flavorjones?

**Comment by seanpdoyle** [comment]
@rafaelfranca I've updated this PR to include two commits:

* https://github.com/rails/rails/pull/43361/changes/7dcb9d236f5095d0b9ad15677ca879b495bff4b1 introduces a new configuration with `:rails_dom_testing` as the default and `:none` as the other supported value.
* https://github.com/rails/rails/pull/43361/changes/2812f7f8c730b8f1b902cd46a34cd8241e5601ac adds first-party support for `Capybara::Minitest::Assertions` through a new `:capybara` configuration value

If the first commit is sufficient to merge, I'm happy to drop the second commit and open it as a separate pull request. While adding first-party support for Capybara is a personal goal of mine, the true goal of this PR is to create an integration seam that enables applications to opt-out of `rails-dom-testing` assertions.

The `:none` value is named to mirror configurations like [config.action_dispatch.show_exceptions = :none](https://guides.rubyonrails.org/configuring.html#actiondispatch-showexceptions), and could be set in tandem with application-side integration with `Capybara::Minitest::Assertions`.

Does re-structuring the diff in this way improve this PR's viability and merge-readiness?
